File: solutions/csharp/rest-api/1/RestApi.cs
using System.Collections.Generic;

using Newtonsoft.Json;

public class RestApi
{
    List<Database> Databases { get; set; }
    public RestApi(string database)
    {
        Databases = JsonConvert.DeserializeObject<List<Database>>(database);
    }

    public string Get(string url, string payload = null)
    {
        string result = "";
        if (url == "/users")
        {
            List<Database> databases = new();
            if (payload != null)
            {
                Users users = JsonConvert.DeserializeObject<Users>(payload);
                foreach (string user in users.users)
                {
                    foreach (Database database in Databases)
                    {
                        if (database.name == user)
                        {
                            databases.Add(database);
                        }
                    }
                }
            }
            result = JsonConvert.SerializeObject(databases);
        }
        return result;
    }

    public string Post(string url, string payload)
    {
        string result = "";
        if (url == "/add")
        {
            User user = JsonConvert.DeserializeObject<User>(payload);
            Database newDatabase = new()
            {
                name = user.user
            };
            Databases.Add(newDatabase);
            result = JsonConvert.SerializeObject(newDatabase);
        }
        else if (url == "/iou")
        {
            IOU iou = JsonConvert.DeserializeObject<IOU>(payload);
            foreach (Database database in Databases)
            {
                if (iou.lender == database.name)
                {
                    if (!database.owes.ContainsKey(iou.borrower))
                    {
                        database.owed_by.Add(iou.borrower, iou.amount);
                    }
                    if (database.owes.ContainsKey(iou.borrower))
                    {
                        database.owes[iou.borrower] -= iou.amount;
                    }
                }
                if (iou.borrower == database.name)
                {
                    if (!database.owed_by.ContainsKey(iou.lender))
                    {
                        database.owes.Add(iou.lender, iou.amount);
                    }
                    if (database.owed_by.ContainsKey(iou.lender))
                    {
                        database.owed_by[iou.lender] -= iou.amount;
                    }
                }
            }

            //Check for zero and negative balances, fixing if needed
            foreach (Database database in Databases)
            {
                List<string> owedItemsToRemove = new();
                List<string> owedByItemsToRemove = new();
                List<string> owedItemsToSwitch = new();
                List<string> owedByItemsToSwitch = new();

                //If items are 0, drop them.
                foreach (var item in database.owes)
                {
                    if (item.Value == 0)
                    {
                        owedItemsToRemove.Add(item.Key);
                    }
                    else if (item.Value < 0)
                    {
                        owedItemsToSwitch.Add(item.Key);
                    }
                }
                foreach (string item in owedItemsToRemove)
                {
                    database.owes.Remove(item);
                }
                foreach (var item in database.owed_by)
                {
                    if (item.Value == 0)
                    {
                        owedByItemsToRemove.Add(item.Key);
                    }
                    else if (item.Value < 0)
                    {
                        owedByItemsToSwitch.Add(item.Key);
                    }
                }
                foreach (string item in owedByItemsToRemove)
                {
                    database.owed_by.Remove(item);
                }

                //If there is a negative balance, switch it from the other side
                foreach (string item in owedItemsToSwitch)
                {
                    database.owed_by.Add(item, -database.owes[item]);
                    database.owes.Remove(item);
                }
                foreach (string item in owedByItemsToSwitch)
                {
                    database.owes.Add(item, -database.owed_by[item]);
                    database.owed_by.Remove(item);
                }
            }


            List<Database> databases = new();
            foreach (Database database in Databases)
            {
                if (database.name == iou.lender || database.name == iou.borrower)
                {
                    databases.Add(database);
                }
            }
            result = JsonConvert.SerializeObject(databases);
        }
        return result;
    }

}

public class IOU
{
    public string lender { get; set; }
    public string borrower { get; set; }
    public int amount { get; set; }
}

public class Users
{
    public List<string> users { get; set; }
}

public class User
{
    public string user { get; set; }
}

public class Database
{
    public Database()
    {
        owes = new();
        owed_by = new();
    }
    public string name { get; set; }
    public SortedDictionary<string, int> owes { get; set; }
    public SortedDictionary<string, int> owed_by { get; set; }
    public int balance
    {
        get
        {
            int balance = 0;
            foreach (KeyValuePair<string, int> item in owes)
            {
                balance -= item.Value;
            }
            foreach (KeyValuePair<string, int> item in owed_by)
            {
                balance += item.Value;
            }
            return balance;
        }
    }
}

File: solutions/java/rest-api/1/src/main/java/RestApi.java
import org.json.JSONArray;
import org.json.JSONObject;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class RestApi {
    private final Map<String, User> usersList;

    public RestApi(User... users) {
        usersList = new HashMap<>();
        for (User user : users) {
            usersList.put(user.name(), user);
        }
    }

    public String get(String url) {
        JSONArray ja = new JSONArray();
        JSONObject out = new JSONObject().put("users", ja);
        for (Map.Entry<String, User> user : usersList.entrySet()) {
            ja.put(getJson(user.getKey()));
        }
        return out.toString();
    }

    public String get(String url, JSONObject payload) {
        JSONArray ja = new JSONArray();
        JSONObject out = new JSONObject().put("users", ja);
        for (Object name : payload.getJSONArray("users")) {
            ja.put(getJson(name.toString()));
        }
        return out.toString();
    }

    public String post(String url, JSONObject payload) {
        switch (url) {
            case "/add":
                add(payload);
                return getJson(payload.getString("user")).toString();
            case "/iou":
                String lender = payload.getString("lender");
                String borrower = payload.getString("borrower");
                double amount = payload.getDouble("amount");
                iou(lender, borrower, amount);

                JSONObject newPayload = new JSONObject().put("users", new JSONArray());
                if (lender.compareTo(borrower) < 0) {
                    newPayload.getJSONArray("users").put(lender).put(borrower);
                } else {
                    newPayload.getJSONArray("users").put(borrower).put(lender);
                }
                return get("/users", newPayload);
        }
        throw new UnsupportedOperationException("Unsupported endpoint");
    }

    private void iou(String lenderName, String borrowerName, double amount) {
        User lender = usersList.get(lenderName);
        User borrower = usersList.get(borrowerName);

        List<Iou> lenderOwes = new LinkedList<>(lender.owes());
        List<Iou> borrowerOwes = new LinkedList<>(borrower.owes());
        borrowerOwes.add(new Iou(lenderName, amount));

        List<Iou> lenderOwedBy = new LinkedList<>(lender.owedBy());
        lenderOwedBy.add(new Iou(borrowerName, amount));
        List<Iou> borrowerOwedBy = new LinkedList<>(borrower.owedBy());

        newBuild(lenderName, lenderOwes, lenderOwedBy);
        newBuild(borrowerName, borrowerOwes, borrowerOwedBy);
    }

    private void newBuild(String name, List<Iou> owes, List<Iou> owedBy) {
        User.Builder builder = User.builder().setName(name);

        outerloop: for (Iou i : owes) {
            for (Iou j : owedBy) {
                if (i.name.equals(j.name)) {
                    if (i.amount > j.amount) {
                        builder.owes(i.name, i.amount - j.amount);
                        continue outerloop;
                    } else {
                        continue outerloop;
                    }
                }
            }
            builder.owes(i.name, i.amount);
        }

        outerloop: for (Iou i : owedBy) {
            for (Iou j : owes) {
                if (i.name.equals(j.name)) {
                    if (i.amount > j.amount) {
                        builder.owedBy(i.name, i.amount - j.amount);
                        continue outerloop;
                    } else {
                        continue outerloop;
                    }
                }
            }
            builder.owedBy(i.name, i.amount);
        }

        usersList.put(name, builder.build());
    }

    private void add(JSONObject payload) {
        String userName = payload.getString("user");
        usersList.put(userName, User.builder().setName(userName).build());
    }

    private JSONObject getJson(String userName) {
        User user = usersList.get(userName);
        JSONObject out = new JSONObject();
        JSONObject owes = new JSONObject();
        JSONObject owedBy = new JSONObject();
        double balance = 0;

        for (Iou owe : user.owes()) {
            owes.put(owe.name, owe.amount);
            balance -= owe.amount;
        }
        for (Iou owed : user.owedBy()) {
            owedBy.put(owed.name, owed.amount);
            balance += owed.amount;
        }

        out.put("name", userName)
           .put("owes", owes)
           .put("owedBy", owedBy)
           .put("balance", balance);
        return out;
    }
}
